File: paper/src/main/java/com/github/aecsocket/calibre/paper/projectile/PaperProjectile.java
package com.github.aecsocket.calibre.paper.projectile;

import com.github.aecsocket.calibre.core.projectile.Projectile;
import com.gitlab.aecsocket.minecommons.core.raycast.Raycast;
import com.gitlab.aecsocket.minecommons.core.scheduler.TaskContext;
import com.gitlab.aecsocket.minecommons.core.vector.cartesian.Vector3;
import com.gitlab.aecsocket.minecommons.paper.PaperUtils;
import com.gitlab.aecsocket.minecommons.paper.raycast.PaperRaycast;
import com.gitlab.aecsocket.sokol.core.tree.TreeNode;
import org.bukkit.World;
import org.bukkit.entity.Entity;
import org.bukkit.entity.EntityType;

import java.util.function.Predicate;

public class PaperProjectile extends Projectile<PaperRaycast.PaperBoundable> {
    private final World world;
    private final Entity shooter;

    private boolean leftShooter;

    public PaperProjectile(TreeNode fullTree, TreeNode localTree, Raycast<PaperRaycast.PaperBoundable> raycast, Vector3 position, Vector3 velocity, World world, Entity shooter) {
        super(fullTree, localTree, raycast, position, velocity);
        this.world = world;
        this.shooter = shooter;
    }

    public World world() { return world; }
    public Entity shooter() { return shooter; }

    @Override
    protected Predicate<PaperRaycast.PaperBoundable> test() {
        return b -> {
            if (b.entity() != null)
                return !b.entity().isDead() && b.entity().getType() != EntityType.DROPPED_ITEM;
            return true;
        };
    }

    @Override
    public void tick(TaskContext ctx) {
        if (PaperUtils.toBukkit(position, world).isChunkLoaded())
            super.tick(ctx);
        else
            ctx.cancel();
    }

    @Override
    protected void noHit(TaskContext ctx, Raycast.Result<PaperRaycast.PaperBoundable> ray, double step, Vector3 oPosition, Vector3 oVelocity) {
        leftShooter = true;
        super.noHit(ctx, ray, step, oPosition, oVelocity);
    }

    @Override
    public void hit(TaskContext ctx, Raycast.Result<PaperRaycast.PaperBoundable> ray, Raycast.Hit<PaperRaycast.PaperBoundable> hit, double step, Vector3 oPosition, Vector3 oVelocity) {
        if (!leftShooter && shooter.equals(hit.hit().entity())) {
            position = hit.out().add(velocity.multiply(1e-7));
            return;
        }
        super.hit(ctx, ray, hit, step, oPosition, oVelocity);
    }
}

File: paper/src/main/kotlin/com/gitlab/aecsocket/calibre/paper/PaperProjectile.kt
package com.gitlab.aecsocket.calibre.paper

import com.github.aecsocket.alexandria.core.bound.Raycast
import com.github.aecsocket.alexandria.core.vector.Vector3
import com.github.aecsocket.alexandria.paper.bound.PaperBoundable
import com.github.aecsocket.alexandria.paper.extension.scheduleDelayed
import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.World
import kotlin.math.abs

private const val EPSILON = 0.1

class PaperProjectile(
    private val plugin: Calibre,
    val world: World,
    source: PaperBoundable?,
    position: Vector3,
    velocity: Vector3,
    physics: Physics,
    fluid: Fluid<PaperBoundable>
) : Projectile<PaperBoundable>(
    plugin.raycast(world),
    source,
    position,
    velocity,
    physics,
    fluid
) {
    override fun computeCollision(dt: Double, result: Raycast.Result<PaperBoundable>) {
        when (result) {
            is Raycast.Result.Hit -> {
                val direction = result.ray.direction
                val epsilon = direction * EPSILON
                position = result.posIn - epsilon

                // apply force

                // enter/exit
                if (!fluid.sameAs(result.hit)) {
                    if (result.hit.fluid =)
                    when (result.hit.fluid) {
                        Material.WATER -> {

                        }
                    }
                }

                // calculate resultant velocity
                if (abs(direction.dot(result.normal)) < 0.2) {
                    // ricochet
                    velocity = Vector3.reflect(velocity, result.normal) * 0.8
                } else {

                }
            }
            is Raycast.Result.Miss -> {
                position = result.position
            }
        }
    }

    fun tick(dt: Double) {
        // todo
        Bukkit.getOnlinePlayers().forEach { if (it.isSneaking) return }

        val from = position

        updatePhysics(dt)

        if (!Location(world, position.x, position.y, position.z).isChunkLoaded)
            return

        computeCollisions(dt)

        val to = position
        val delta = to - from

        val direction = delta.normalized
        val effector = plugin.effectors.world(world)
        val particles = plugin.settings.trailParticle.map { it.copy(size = velocity) }

        var step = 0.0
        while (step < delta.length) {
            particles.forEach { effector.showParticle(it, from + direction * step) }
            step += 1.0
        }

        plugin.scheduleDelayed(1) { tick(0.05) }
    }
}
